src/estimators/stratified_sketch.py

```diff
@@ -0,0 +1,242 @@
+  """Merge and estimate two StratifiedSketch."""
+
+  def __init__(self, sketch_operator):
+    """Create an estimator for two Stratified sketches.
+
+    Args:
+      sketch_operator: an object that have union, intersection, and
+        difference methods for two sketches.
+    """
+
+    self.sketch_union = sketch_operator.union
+    self.sketch_difference = sketch_operator.difference
+    self.sketch_intersection = sketch_operator.intersection
+
+  def __call__(self, this, that):
+    merged = self.merge_sketches(this, that)
+    return merged
+
+  def merge_sketches(self, this, that):
+    """Merge two StratifiedSketch.
+
+       Given 2 sketches A and B:
+       Merged(k) = (A(k) & B(0)) U (A(k-1) & B(1)) ... U (A(0) & B(k))
+         where
+         A(k) & B(0) =  A(k) - (A(k) & B(1+))
+         B(k) & A(0) =  B(k) - (B(k) & A(1+))
+    Args:
+      this: one of the two StratifiedSketch to be merged.
+      that: the other StratifiedSketch to be merged.
+
+    Returns:
+      A merged StratifiedSketch from the input.
+    """
+    assert isinstance(this, StratifiedSketch)
+    this.assert_compatible(that)
+    max_freq = this.max_freq
+    merged_sketch = copy.deepcopy(this)
+
+    for k in range(max_freq - 1):
+      # Calculate A(k) & B(0) = A(k) - (A(k) & B(1+))
+      merged = self.sketch_difference(
+          this.sketches[str(k + 1)], self.sketch_intersection(
+              this.sketches[str(k + 1)], that.sketches[ONE_PLUS]))
+
+      # Calculate A(0) & B(k) = B(k) - (B(k) & A(1+))
+      merged = self.sketch_union(
+          merged, self.sketch_difference(
+              that.sketches[str(k + 1)], self.sketch_intersection(
+                  this.sketches[ONE_PLUS], that.sketches[str(k + 1)])))
+
+      # Calculate A(i) & B(k-i)
+      for i in range(k):
+        merged = self.sketch_union(
+            merged, self.sketch_intersection(
+                this.sketches[str(i + 1)], that.sketches[str(k - i)]))
+      merged_sketch.sketches[str(k + 1)] = merged
+
+    # Calculate Merged(max_freq)
+    merged = this.sketches[str(max_freq)]
+    rest = that.sketches[ONE_PLUS]
+    for k in range(max_freq - 1):
+      merged = self.sketch_union(
+          merged,
+          self.sketch_intersection(this.sketches[str(max_freq - k - 1)],
+                                   rest))
+      rest = self.sketch_difference(rest, that.sketches[str(k + 1)])
+    merged = self.sketch_union(
+        merged,
+        self.sketch_difference(
+            that.sketches[str(max_freq)],
+            self.sketch_intersection(
+                that.sketches[str(max_freq)],
+                this.sketches[ONE_PLUS])))
+    merged_sketch.sketches[str(max_freq)] = merged
+
+    # Calculate Merged(1+)
+    merged_one_plus = None
+    for k in range(max_freq):
+      merged_one_plus = self.sketch_union(merged_one_plus,
+                                          merged_sketch.sketches[str(k + 1)])
+    merged_sketch.sketches[ONE_PLUS] = merged_one_plus
```

Comment:
Is it the case that
  merged_one_plus = self.sketch_union(this.sketches[ONE_PLUS], that.sketches[ONE_PLUS])?

Reply:
@huangxichen1 I aggree with Matthew, that it should be the case. What do you think?